mfe.go

```diff
@@ -0,0 +1,1092 @@
+		encodedSequence[i] = nucleotideRuneMap[sequence[i]]
+	}
+
+	return encodedSequence
+}
+
+/**
+* Returns a map that encodes a base pair to its numerical representation.
+* The map is:
+*    _  A  C  G  U
+*	_ {0, 0, 0, 0, 0}
+*	A {0, 0, 0, 0, 5}
+*	C {0, 0, 0, 1, 0}
+*	G {0, 0, 2, 0, 3}
+*	U {0, 6, 0, 4, 0}
+* For example, map['A']['U'] is 5.
+* The encoded numerical representation of a base pair carries no meaning in
+* itself except for where to find the relevent energy contributions of the base
+* pair in the matrices of the energy paramaters (found in `energy_params.go`).
+* Thus, any change to this map must be reflected in the energy
+* parameters matrices in `energy_params.go`, and vice versa (see
+* `energy_params.md` for more information.)
+ */
+func basePairEncodedTypeMap() map[byte]map[byte]int {
+	nucelotideAEncodedTypeMap := map[byte]int{'U': 5}
+	nucelotideCEncodedTypeMap := map[byte]int{'G': 1}
+	nucelotideGEncodedTypeMap := map[byte]int{'C': 2, 'U': 3}
+	nucelotideUEncodedTypeMap := map[byte]int{'A': 6, 'G': 4}
+
+	basePairMap := map[byte]map[byte]int{
+		'A': nucelotideAEncodedTypeMap,
+		'C': nucelotideCEncodedTypeMap,
+		'G': nucelotideGEncodedTypeMap,
+		'U': nucelotideUEncodedTypeMap,
+	}
+
+	return basePairMap
+}
+
+// Contains all the energy parameters needed for the free energy calculations
+type energyParams struct {
+	stackingPair [nbPairs + 1][nbPairs + 1]int
+	hairpinLoop  [31]int
+	bulge        [maxLoop + 1]int
+	/* This field was previous called internal_loop, but has been renamed to
+	interior loop to remain consistent with the names of other interior loops */
+	interiorLoop                              [maxLoop + 1]int
+	mismatchExteriorLoop                      [nbPairs + 1][5][5]int
+	mismatchInteriorLoop                      [nbPairs + 1][5][5]int
+	mismatch1xnInteriorLoop                   [nbPairs + 1][5][5]int
+	mismatch2x3InteriorLoop                   [nbPairs + 1][5][5]int
+	mismatchHairpinLoop                       [nbPairs + 1][5][5]int
+	mismatchMultiLoop                         [nbPairs + 1][5][5]int
+	dangle5                                   [nbPairs + 1][5]int
+	dangle3                                   [nbPairs + 1][5]int
+	interior1x1Loop                           [nbPairs + 1][nbPairs + 1][5][5]int
+	interior2x1Loop                           [nbPairs + 1][nbPairs + 1][5][5][5]int
+	interior2x2Loop                           [nbPairs + 1][nbPairs + 1][5][5][5][5]int
+	ninio                                     [5]int
+	lxc                                       float64
+	MLbase, MLclosing, TerminalAU, DuplexInit int
+	MLintern                                  [nbPairs + 1]int
+	Tetraloops                                string
+	Tetraloop                                 [200]int
+	Triloops                                  string
+	Triloop                                   [40]int
+	Hexaloops                                 string
+	Hexaloop                                  [40]int
```

Comment:
Why are some attributes private and some public?

Reply:
Their field names didn't follow a specific convention with regards to capitalisation. So the public fields just happen to be field names which they capitalised. Good catch - will rename to make them all private.